lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2950,6 +2950,41 @@ ResolveWitnessResult ConformanceChecker::resolveTypeWitnessViaLookup(
   return ResolveWitnessResult::ExplicitFailed;
 }
 
+static bool associatedTypesAreSameEquivalenceClass(AssociatedTypeDecl *a,
+                                                   AssociatedTypeDecl *b,
+                                                   DeclContext *DC) {
+  auto &C = a->getASTContext();
+  auto *M = DC->getParentModule();
+
+  if (a == b)
+    return true;
+  
+  // Two different assoc types in the same protocol are never equivalent.
+  if (a->getProtocol() == b->getProtocol())
+    return false;
+  
+  // Two different-named assoc types are (currently) never equivalent.
+  if (a->getName() != b->getName())
```

Comment:
The two early exits above, plus the fact that two nested types with the same name and base are automatically made equivalent by the archetype builder, mean that this whole function is effectively the same as `a->getName() == b->getName()` (and an expensive one at that).

I think we should either go all the way to actually comparing the dependent types without these restrictions, or just stick with the name comparison.